src/dbnode/persist/fs/migration/migration.go

```diff
@@ -0,0 +1,101 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package migration
+
+import (
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/storage"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// Migration interface is implemented by tasks that wish to perform a data migration
+// on a fileset. Typically involves updating files in a fileset that were created by
+// a previous version of the database client
+type Migration interface {
+
+	// Run is the set of steps to successfully complete a migration
+	Run() error
+}
+
+var _ Migration = &ToVersion1_1{}
+
+// ToVersion1_1 is an object responsible for migrating a fileset to version 1.1
+type ToVersion1_1 struct {
+	newMergerFn fs.NewMergerFn
+	infoFile    fs.ReadInfoFileResult
+	shard       uint32
+	namespace   namespace.Metadata
+	opts        storage.Options
+}
+
+func NewToVersion1_1(infoFile fs.ReadInfoFileResult, shard uint32, namespace namespace.Metadata, opts storage.Options) Migration {
+	return ToVersion1_1{
+		newMergerFn: fs.NewMerger,
+		infoFile:    infoFile,
+		shard:       shard,
+		namespace:   namespace,
+		opts:        opts,
+	}
+}
+
+// Run executes the steps to bring a fileset to Version 1.1
+func (v ToVersion1_1) Run() error {
+	reader, err := fs.NewReader(v.opts.BytesPool(), v.opts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		return err
+	}
+
+	merger := v.newMergerFn(reader, v.opts.DatabaseBlockOptions().DatabaseBlockAllocSize(),
+		v.opts.SegmentReaderPool(), v.opts.MultiReaderIteratorPool(),
+		v.opts.IdentifierPool(), v.opts.EncoderPool(), v.opts.ContextPool(), v.namespace.Options())
+
+	volIndex := v.infoFile.Info.VolumeIndex
+	fsID := fs.FileSetFileIdentifier{
+		Namespace:   v.namespace.ID(),
+		Shard:       v.shard,
+		BlockStart:  xtime.FromNanoseconds(v.infoFile.Info.BlockStart),
+		VolumeIndex: volIndex,
+	}
+
+	nsCtx := namespace.NewContextFrom(v.namespace)
+
+	flushPersist, err := v.opts.PersistManager().StartFlushPersist()
+	if err != nil {
+		return err
+	}
+
+	// Intentionally use a noop merger here as we simply want to rewrite the same files with the current encoder which
+	// will generate index files with the entry level checksums.
+	close, err := merger.Merge(fsID, storage.NewNoopMergeWith(), volIndex+1, flushPersist, nsCtx,
+		&persist.NoOpColdFlushNamespace{})
+	if err != nil {
+		return err
+	}
+
+	if err := close(); err != nil {
```

Comment:
Bonus question: I want to add an option to `Merge` to delete the original fileset once done. Is there any issue with me having a code path that calls close in merge and then sweeps the previous fileset away?

Reply:
I think that's fine, perhaps we make it a separate method call to `fs.Merger` i.e. `MergeAndCleanup(...)` and have both `Merge(..)` and `MergeAndCleanup(....)` call the same underlying code but have the AndCleanup version delete the old file too?

Reply:
I don't see any issue doing this during bootstrap but I'm not sure we should do this in more places than one. It might be confusing going forward since I think we only do cleanup logic in the cleanup manager to make things easier to reason about.

Also, if we do proceed w/ this I think we'd want to restrict it during bootstrap only? Since I think there are issues if we perform fileset deletion after merging/writing to disk in the TSDB cold flush path as we may be serving reads from the data? 

@robskillington thoughts on this?

Reply:
FWIW, migrations will only be run during bootstrap and block until they complete. Additionally, we can expose some methods on the CleanupManager to at least isolate all the logic in one place.

Reply:
So the main reason we wanted to do this in the bootstrapper @notbdu is that we are going to have to migrate all files. That means literally every single file will need a cleanup and if the cleanup manager doesn't run till the end, we will 2x the amount of data on disk during the migration. This will likely mean ppl run out of disk space before being able to finish the bootstrap.

That's why we need to cleanup the old files during bootstrap. 

We still need the cleanup manager to do it's thing since sometimes this could crash half way through, so we do kind of need it to happen in two places.

So it's made a little bit more complicated by the fact that some folks will need to migrate every single file (hence 2xing size on disk total during the migration, unless we cleanup along the way).